### PR TITLE
Added @softscope macro for ease of use

### PR DESCRIPTION
**Pull request by dawbarton**
The title says it all really. Thanks for making this package.

### CODE REVIEW

**Review by stevengj — DISMISSED** [review]

**Comment by stevengj** [comment]
LGTM, once the merge conflict is fixed.

**Comment by dawbarton** [comment]
Great, I'll have a look at it tomorrow. (I might switch around how the escaping is handled since I'm not quite happy with it.)

File: src/SoftGlobalScope.jl
Comment (stevengj) [inline src/SoftGlobalScope.jl]: Isn't `__module__` new in Julia 0.7 (JuliaLang/julia#22064)?   SoftScope.jl has to support 0.6, where I guess you would use `current_module()`.

**Comment by stevengj** [comment]
Needs a test.

**Comment by dawbarton** [comment]
Working on it; consider this a WIP for the moment.

**Comment by dawbarton** [comment]
OK, all done. Should be v0.6+ compatible and has tests.

**Comment by Datseris** [comment]
Sorry to invade this PR, but I think it is better to ask here than to open a new issue.
In https://github.com/JuliaLang/julia/issues/28789 it was proposed that the following is possible: 
```
julia> a = 0                                                                
0                                                                           
                                                                            
julia> for i = 1:10                                                         
         a += i                                                             
       end                                                                  
ERROR: UndefVarError: a not defined                                         
Stacktrace:                                                                 
 [1] top-level scope at .\REPL[2]:2 [inlined]                               
 [2] top-level scope at .\none:0                                            
                                                                            
julia> using SoftGlobalScope                                                
[ Info: Precompiling SoftGlobalScope [363c7d7e-a618-11e8-01c4-4f22c151e122] 
                                                                            
julia> for i = 1:10                                                         
         a += i                                                             
       end                                                                  
                                                                            
julia> a                                                                    
55  
```
i.e. the `@softscope` macro introduced here is applied everywhere by default. Is this still possible to be done?

**Comment by dawbarton** [comment]
Hi @Datseris, see #2 for that request. I looked into it further this afternoon and I have a few suggestions for a way forward (I'll write up in the next few days). In the meantime, I've updated the gist at https://gist.github.com/dawbarton/0388715fb56fb5cd05e0e4b12c322815 which implements my first attempt at the REPL hooks. Note the comment in the code if you use Revise! (Don't try putting the code in your `startup.jl`; it won't work there at the moment.)

**Comment by stevengj** [comment]
I modified the test to make sure it works with a variable that is defined in the `@softscope` block.   I don't think it will at the moment— this is why `softscope_include_string` has to evaluate one statement at a time.

**Comment by dawbarton** [comment]
Two can play at that game! :smile: 

Try this with `softscope_include_string` -
```
softscope_include_string(Main, "if true ; ab = 0; for i = 1:10; ab += i ; end ; end")
```

I think we'll need to recurse into blocks and record new assignments rather than just trying to execute one block at a time (which will obviously fail here). I've got a stack of exam marking to do by close of play tomorrow so I won't be able to have a look before then but I'll have a go at it after that if you haven't already had a look at it.

**Comment by stevengj** [comment]
@dawbarton, yes, that occurred to me.  It's not high on my priority list at the moment since I'm guessing that defining new global variables in blocks is not so common in interactive code, but we can definitely catch some common cases like `x = ....` assignments in `begin` and `if` blocks.

However, without that feature I suspect that `@softscope` is much less useful, since `@softscope begin ... end` is probably the most useful formulation.

(It will be possible to fool `softscope` with things like `@eval`, but that's even further down on my list of priorities.)

**Comment by dawbarton** [comment]
This should now all work correctly with #4 merged.

**Pull request merged**
